File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/standings/StandingsDataSupport.kt
package com.pillyliu.pinprofandroid.standings

import com.pillyliu.pinprofandroid.data.parseCsv
import java.text.NumberFormat
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

internal const val STANDINGS_CSV_URL = "https://pillyliu.com/pinball/data/LPL_Standings.csv"

internal fun buildStandings(rows: List<StandingsCsvRow>, selectedSeason: Int?): List<Standing> {
    if (selectedSeason == null) return emptyList()
    val seasonRows = rows.filter { it.season == selectedSeason }
    if (seasonRows.isEmpty()) return emptyList()

    val mapped = seasonRows.map {
        Standing(
            rawPlayer = it.player,
            seasonTotal = it.total,
            eligible = it.eligible,
            nights = it.nights,
            banks = it.banks,
        )
    }

    val hasRankForAll = seasonRows.all { it.rank != null }
    if (hasRankForAll) {
        val rankMap = seasonRows.associate { it.player to (it.rank ?: Int.MAX_VALUE) }
        return mapped.sortedBy { rankMap[it.rawPlayer] ?: Int.MAX_VALUE }
    }

    return mapped.sortedByDescending { it.seasonTotal }
}

internal fun parseStandings(text: String): List<StandingsCsvRow> {
    val table = parseCsv(text)
    if (table.isEmpty()) return emptyList()

    val headers = table[0].map { normalizeStandingsHeader(it) }
    val required = listOf(
        "season", "player", "total", "bank_1", "bank_2", "bank_3", "bank_4",
        "bank_5", "bank_6", "bank_7", "bank_8",
    )
    if (required.any { it !in headers }) {
        throw IllegalStateException("Standings CSV missing required columns")
    }

    return table.drop(1).mapNotNull { row ->
        if (row.size != headers.size) return@mapNotNull null
        val dict = headers.zip(row).toMap()

        val season = coerceStandingsSeason(dict["season"].orEmpty())
        val player = dict["player"].orEmpty().trim()
        if (season <= 0 || player.isBlank()) return@mapNotNull null

        StandingsCsvRow(
            season = season,
            player = player,
            total = dict["total"].orEmpty().toDoubleOrNull() ?: 0.0,
            rank = dict["rank"].orEmpty().trim().toIntOrNull(),
            eligible = dict["eligible"].orEmpty().trim(),
            nights = dict["nights"].orEmpty().trim(),
            banks = (1..8).map { i -> dict["bank_$i"].orEmpty().toDoubleOrNull() ?: 0.0 },
        )
    }
}

internal fun formatStandingsValue(value: Double): String = NumberFormat.getIntegerInstance().format(value.toLong())

internal fun formatStandingsUpdatedAt(epochMs: Long): String {
    return SimpleDateFormat("M/d/yy h:mm a", Locale.getDefault()).format(Date(epochMs))
}

private fun normalizeStandingsHeader(header: String): String {
    return header.replace("\uFEFF", "").trim().lowercase()
}

private fun coerceStandingsSeason(value: String): Int {
    val trimmed = value.trim()
    val digits = trimmed.filter { it.isDigit() }
    return digits.toIntOrNull() ?: trimmed.toIntOrNull() ?: 0
}

File: Pinball App 2/Pinball App 2/standings/StandingsDataSupport.swift
import Foundation

enum StandingsCSVLoader {
    static let defaultPath = hostedLeagueStandingsPath

    static func parse(text: String) throws -> [StandingsCSVRow] {
        let table = parseCSVRows(text)
        guard !table.isEmpty else { return [] }

        let headers = table[0].map { normalizeCSVHeader($0) }
        let required = [
            "season", "player", "total", "bank_1", "bank_2", "bank_3", "bank_4",
            "bank_5", "bank_6", "bank_7", "bank_8"
        ]

        for name in required where !headers.contains(name) {
            throw StandingsCSVError.missingColumn(name)
        }

        return table.dropFirst().compactMap { row in
            guard row.count == headers.count else { return nil }

            let dict = Dictionary(uniqueKeysWithValues: zip(headers, row))

            let season = coerceSeasonNumber(dict["season"] ?? "")
            let player = (dict["player"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            let total = Double(dict["total"] ?? "") ?? 0
            let rank = Int((dict["rank"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines))
            let eligible = (dict["eligible"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            let nights = (dict["nights"] ?? "").trimmingCharacters(in: .whitespacesAndNewlines)

            let banks = (1...8).map { index in
                Double(dict["bank_\(index)"] ?? "") ?? 0
            }

            guard season > 0, !player.isEmpty else { return nil }

            return StandingsCSVRow(
                season: season,
                player: player,
                total: total,
                rank: rank,
                eligible: eligible,
                nights: nights,
                banks: banks
            )
        }
    }
}

enum StandingsCSVError: LocalizedError {
    case missingColumn(String)
    case network(String)

    var errorDescription: String? {
        switch self {
        case .missingColumn(let column):
            return "Standings CSV missing column: \(column)"
        case .network(let message):
            return message
        }
    }
}

func formatStandingsRounded(_ value: Double) -> String {
    Int(value.rounded()).formatted()
}
